### DISCUSSION TITLE
how can I get Type.eStorage

### DISCUSSION THREAD

**Discussion by keche-master** [eclipse-emf/org.eclipse.emf#39] [Q&A] [answered]
how can I get Type.eStorage

**Comment by merks** [comment, marked as answer]
Cast to InternalEObject and use org.eclipse.emf.ecore.InternalEObject.eStore().  For most implementations this will always be null.  I wonder what you are trying to do?

**Reply by keche-master** [reply]
I import the parameters in one of these elements like this 
<packagedElement xmi:type="uml:Class" xmi:id="_LTigFEqLEe-YQOzO5UKUfw" name="Class">
            
    <ownedAttribute xmi:type="uml:Property" xmi:id="_LTigFUqLEe-YQOzO5UKUfw" name="Attribute" visibility="public" isUnique="false">
                  
        <type xmi:type="uml:PrimitiveType" href="http://www.omg.org/spec/UML/20110701/PrimitiveTypes.xmi#String"/>
                
    </ownedAttribute>
          
 </packagedElement> 
, and I need to get the java type for this element called Attribute

**Reply by keche-master** [reply]
When I debug, I found that there is a property in eStorage that can be taken directly to String

**Reply by merks** [reply]
Sorry, I misread it as eStore not eStorage.   

I assume you have a UML2 Property instance with a getType method and you should cast to that and call getType.   You should look at the UML2 APIs of these instances...

**Reply by keche-master** [reply]
I'm currently using the EObject type to get the EAttribute, which is a Property element, but I haven't found an API that can get the type. getType() after getting it, check the properties through debug, and find that there is the data I need under eStorage, but I can't get the data....

**Reply by merks** [reply]
For any EObject x, you can use x.eClass().getEAllStructureFeatures() to get all the features.  Then you can call x.eGet(f) for each such structural feature.  I wonder what is the implementation class of the instance?  There isn't a PropertyImpl or PrimitiveTypeImpl involved?   Something like x.eGet(x.eClass().getEStructuralFeature("type")) would return the value of the type feature if the object has a feature by that name.  EObject.eGet is how you should get values out of an EObject, assuming you don't have generated API for that object.

**Reply by keche-master** [reply]
![image](https://github.com/user-attachments/assets/84476474-ca92-42ac-be93-5b34d3b04b1f)
The data I want is in this eStorage, and I found that the getField(int field) method can get some data, but after taking it out, it is almost the same as the data I got by calling the eAdapters method directly

**Reply by keche-master** [reply]
The implementation class of the instance is PropertyImpl

**Reply by merks** [reply]
So can't you cast it to org.eclipse.uml2.Property and call getType()?  The only generic API to get data values from an EObject is via eGet, but I've already said that...

**Reply by keche-master** [reply]
Thanks for the answer, I will try to solve my problem using this method
